clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java

```diff
@@ -3263,6 +3282,95 @@ public void testListConsumerGroupOffsets() throws Exception {
         }
     }
 
+    @Test
+    public void testBatchedListConsumerGroupOffsets() throws Exception {
+        Cluster cluster = mockCluster(1, 0);
+        Time time = new MockTime();
+        Map<String, ListConsumerGroupOffsetsSpec> groupSpecs = batchedListConsumerGroupOffsetsSpec();
+
+        try (AdminClientUnitTestEnv env = new AdminClientUnitTestEnv(time, cluster, AdminClientConfig.RETRIES_CONFIG, "0")) {
+            env.kafkaClient().setNodeApiVersions(NodeApiVersions.create());
+            env.kafkaClient().prepareResponse(prepareBatchedFindCoordinatorResponse(Errors.NONE, env.cluster().controller(), groupSpecs.keySet()));
+
+            ListConsumerGroupOffsetsResult result = env.adminClient().listConsumerGroupOffsets(groupSpecs, new ListConsumerGroupOffsetsOptions());
+            sendOffsetFetchResponse(env.kafkaClient(), groupSpecs, true);
+
+            verifyListOffsetsForMultipleGroups(groupSpecs, result);
+        }
+    }
+
+    @Test
+    public void testBatchedListConsumerGroupOffsetsWithNoBatchingSupport() throws Exception {
+        Cluster cluster = mockCluster(1, 0);
+        Time time = new MockTime();
+        Map<String, ListConsumerGroupOffsetsSpec> groupSpecs = batchedListConsumerGroupOffsetsSpec();
+
+        ApiVersion findCoordinatorV3 = new ApiVersion()
```

Comment:
Should we also add a test where the FindCoordinator supports v4 and OffsetFetch does not support v8? I think that this could happen in a partially upgraded cluster.

Reply:
I wanted to do this yesterday, but it seemed too hard to do with the current client APIs for MockClient because we don't seem to use request version unless we prepare the response in advance. For the sequence of unbatched calls with multiple group ids, it is difficult to prepare the response in advance when there are multiple groups since they can appear in any order. So basically my test worked, but it didn't actually give partial batching I was expecting. Do you think we really need to do this?

Reply:
It would be great to have it because this is the kind of thing that we will forget if we refactor this code in the future. Is the order of the groups important in the response? If it is not, we could prepare a response with the group we expect. If the concern is about verifying the request, we could at minimum ensures that the expected group are in the request without caring about their order. Would this work?

Reply:
@dajac I figured out the test, one failed response prepared in advance is sufficient to force switching to non-batched mode. Thereafter the requests can be sent later where we match groups from the request as we do in the other test.